File: backend/app/elo.py
"""Elo-style knowledge score.

Each user has one rating per format (starting at 1000). Every quiz question
acts as an opponent whose rating is derived from the post's difficulty:
post_difficulty 1 -> 800, 2 -> 1000, 3 -> 1200.

Standard Elo update: R' = R + K * (S - E) where S is 1 (correct) or 0 (wrong)
and E = 1 / (1 + 10^((Q - R) / 400)) is the expected score against a question
rated Q. Wrong answers therefore always cost points, so guessing has a cost.

K is 32 for a user's first 30 answers in a format (ratings converge quickly
while provisional) and 16 afterwards (stable scores move slowly). Ratings are
floored at 100 so a losing streak can never produce absurd negative numbers.

The global score is the average of the user's per-format ratings, over the
formats where the user has answered at least one question.
"""

from sqlalchemy.orm import Session

from .models import UserElo

START_RATING = 1000.0
FLOOR_RATING = 100.0
K_PROVISIONAL = 32.0
K_STABLE = 16.0
PROVISIONAL_ANSWERS = 30

DIFFICULTY_RATING = {1: 800.0, 2: 1000.0, 3: 1200.0}


def question_rating(post_difficulty) -> float:
    return DIFFICULTY_RATING.get(post_difficulty, DIFFICULTY_RATING[2])


def expected_score(user_rating: float, q_rating: float) -> float:
    return 1.0 / (1.0 + 10.0 ** ((q_rating - user_rating) / 400.0))


def get_or_create_elo(db: Session, user_id: int, fmt: str) -> UserElo:
    row = db.query(UserElo).filter(
        UserElo.user_id == user_id, UserElo.format == fmt
    ).first()
    if row is None:
        row = UserElo(user_id=user_id, format=fmt, rating=START_RATING, answered_count=0)
        db.add(row)
        db.flush()
    return row


def apply_answer(db: Session, user_id: int, fmt: str, post_difficulty, correct: bool) -> tuple[UserElo, float]:
    """Update the user's per-format rating for one first-time answer.

    Returns the UserElo row and the rating delta. Caller commits.
    """
    row = get_or_create_elo(db, user_id, fmt)
    k = K_PROVISIONAL if row.answered_count < PROVISIONAL_ANSWERS else K_STABLE
    expected = expected_score(row.rating, question_rating(post_difficulty))
    actual = 1.0 if correct else 0.0
    delta = k * (actual - expected)
    row.rating = max(FLOOR_RATING, row.rating + delta)
    row.answered_count += 1
    return row, delta


def global_rating(db: Session, user_id: int) -> int | None:
    """Average of per-format ratings; None until the user has answered a quiz."""
    rows = db.query(UserElo).filter(UserElo.user_id == user_id).all()
    if not rows:
        return None
    return round(sum(r.rating for r in rows) / len(rows))


def format_ratings(db: Session, user_id: int) -> dict[str, dict]:
    rows = db.query(UserElo).filter(UserElo.user_id == user_id).all()
    return {
        r.format: {"rating": round(r.rating), "answered_count": r.answered_count}
        for r in rows
    }

File: frontend/src/lib/train/elo.ts
// Marathon Elo simulator + adaptive difficulty picker. Ported verbatim from the
// mobile app (mobile/src/lib/train/elo.ts) so the guest simulation matches.
//
// The AUTHORITATIVE knowledge rating is the SERVER's per-format Elo
// (backend/app/elo.py, persisted in UserElo). This module simulates a
// MARATHON-ONLY rating purely for the UI while the marathon backend does not
// exist yet; the number it produces is not stored anywhere and is not the real
// rating. Once the marathon endpoints land it must be reconciled with / replaced
// by the server's computation.
//
// The core math mirrors the server (same expected-score formula, same K-factors,
// same difficulty ratings, same "wrong always costs the full base" philosophy)
// so the simulated number behaves like the real one. A time bonus is layered on
// top for the marathon's feel -- it only ever ADDS to a correct gain.
//
// Pure module: no React, no I/O, no randomness except the explicit weighted
// difficulty pick. Keep it that way so it stays unit-testable.

// --- Tunable constants (single source of truth for the marathon math). ---
export const START_ELO = 1000
export const ELO_FLOOR = 100
export const K_FAST = 32 // first 30 scored answers (provisional, converges fast)
export const K_SLOW = 16 // after 30 answers (stable, moves slowly)
export const DIFFICULTY_RATING: Record<number, number> = { 1: 800, 2: 1000, 3: 1200 }
export const FAST_MS = 3000 // at/under this, full time bonus
export const SLOW_MS = 15000 // at/over this, no time bonus
export const TIME_BONUS_MAX = 0.5 // fastest correct answer earns up to +50% of the base gain

// Standard Elo expected score: probability R beats an opponent rated Q.
export function expectedScore(R: number, Q: number): number {
  return 1 / (1 + 10 ** ((Q - R) / 400))
}

// K is larger while the rating is provisional so early answers move it quickly.
export function kFactor(answeredCount: number): number {
  return answeredCount < 30 ? K_FAST : K_SLOW
}

// 1 at/under FAST_MS, 0 at/over SLOW_MS, linear in between (clamped 0..1).
export function timeFactor(answerMs: number): number {
  if (answerMs <= FAST_MS) return 1
  if (answerMs >= SLOW_MS) return 0
  return (SLOW_MS - answerMs) / (SLOW_MS - FAST_MS)
}

// Signed, rounded rating change for one answer. Time only sweetens a correct
// gain; a wrong answer always costs the full base, so guessing has a real cost
// (matching the server philosophy).
export function computeDelta(params: {
  R: number
  difficulty: number
  correct: boolean
  answerMs: number
  answeredCount: number
}): number {
  const { R, difficulty, correct, answerMs, answeredCount } = params
  const Q = DIFFICULTY_RATING[difficulty] ?? DIFFICULTY_RATING[2]
  const base = kFactor(answeredCount) * ((correct ? 1 : 0) - expectedScore(R, Q))
  const delta = correct ? base * (1 + TIME_BONUS_MAX * timeFactor(answerMs)) : base
  return Math.round(delta)
}

// Apply a delta to a rating, floored so a losing streak cannot go absurdly low.
export function applyDelta(R: number, delta: number): number {
  return Math.max(ELO_FLOOR, Math.round(R + delta))
}

// Weighted random pick from a { value: weight } map.
function weightedPick(weights: Record<number, number>): number {
  const entries = Object.entries(weights)
  const total = entries.reduce((sum, [, w]) => sum + w, 0)
  let roll = Math.random() * total
  for (const [value, w] of entries) {
    roll -= w
    if (roll <= 0) return Number(value)
  }
  return Number(entries[entries.length - 1][0])
}

// Map current Elo to a preferred difficulty with light variety, so the marathon
// is not deterministic. The intent is to keep expectedScore near ~0.5 (questions
// matched to skill), which is also where the rating actually moves.
export function pickDifficulty(R: number): number {
  if (R < 900) return weightedPick({ 1: 0.8, 2: 0.2 })
  if (R <= 1150) return weightedPick({ 1: 0.2, 2: 0.6, 3: 0.2 })
  return weightedPick({ 2: 0.25, 3: 0.75 })
}
